components/raftstore-v2/src/operation/command/admin/raft_log_gc.rs

```diff
@@ -0,0 +1,279 @@
+        // TODO: move this into a function
+        let term = self.raft_group().raft.raft_log.term(compact_idx).unwrap();
+
+        let mut req = new_admin_request(self.region_id(), self.peer().clone());
+        let mut admin = AdminRequest::default();
+        admin.set_cmd_type(AdminCmdType::CompactLog);
+        admin.mut_compact_log().set_compact_index(compact_idx);
+        admin.mut_compact_log().set_compact_term(term);
+        req.set_admin_request(admin);
+
+        let (ch, _) = CmdResChannel::pair();
+        self.on_admin_command(store_ctx, req, ch);
+
+        self.reset_skip_raft_log_gc_ticks();
+    }
+}
+
+#[derive(Debug)]
+pub struct CompactLogResult {
+    compact_index: u64,
+    compact_term: u64,
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    pub fn propose_compact_log<T>(
+        &mut self,
+        store_ctx: &mut StoreContext<EK, ER, T>,
+        req: RaftCmdRequest,
+    ) -> Result<u64> {
+        let compact_log = req.get_admin_request().get_compact_log();
+        // TODO: add unit tests to cover all the message integrity checks.
+        if compact_log.get_compact_term() == 0 {
+            info!(
+                self.logger,
+                "compact term missing, skip";
+                "command" => ?compact_log
+            );
+            // old format compact log command, safe to ignore.
+            return Err(box_err!(
+                "command format is outdated, please upgrade leader"
+            ));
+        }
+
+        let data = req.write_to_bytes().unwrap();
+        self.propose_with_ctx(store_ctx, data, vec![])
+    }
+}
+
+impl<EK: KvEngine, R: ApplyResReporter> Apply<EK, R> {
+    pub fn apply_compact_log(
+        &mut self,
+        req: &AdminRequest,
+        _log_index: u64,
+    ) -> Result<(AdminResponse, AdminCmdResult)> {
+        Ok((
+            AdminResponse::default(),
+            AdminCmdResult::CompactLog(CompactLogResult {
+                compact_index: req.get_compact_log().get_compact_index(),
+                compact_term: req.get_compact_log().get_compact_term(),
+            }),
+        ))
+    }
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    pub fn on_apply_res_compact_log<T>(
+        &mut self,
+        store_ctx: &mut StoreContext<EK, ER, T>,
+        mut res: CompactLogResult,
+    ) {
+        let first_index = self.entry_storage().first_index();
+        if res.compact_index <= first_index {
+            debug!(
+                self.logger,
+                "compact index <= first index, no need to compact";
+                "compact_index" => res.compact_index,
+                "first_index" => first_index,
+            );
+            return;
+        }
+        // TODO: check is_merging
+        // compact failure is safe to be omitted, no need to assert.
+        if res.compact_index <= self.entry_storage().truncated_index() {
+            return;
+        }
+        res.compact_index = std::cmp::min(
```

Comment:
What if those writes are flushed latter? I think you need a queue like `split_trace` to trace the compacted index and gc them after applied index is advanced.

Reply:
Rethink about it, the simplest way to do this is to do raft log gc whenever there is data flushed or raft log gc applied. The min(truncated_index, admin.flushed) is chosen so there doesn't need to be queue.

Reply:
How about trigger gc in peer::on_data_flushed, and do the actual gc in store async writer. I think avoiding sync write in apply pool is better in the long run.

Reply:
> trigger gc in peer::on_data_flushed, and do the actual gc in store async writer.

That's amost what I mean. But it can't be done in `on_data_flushed`, it can only delete old logs when `persisted_applied` is advanced.

Reply:
So in the latest version I check for persisted_applied in `maybe_compact_log_from_engine`.